Add RPM packaging support.

mr-salty opened the pull request.

mr-salty committed:
Add RPM packaging support.

katzj commented on pkg/el/google-fluentd-catch-all-config.spec:
```diff
@@ -0,0 +1,40 @@
+Name: google-fluentd-catch-all-config
+Version: 0.1
+Release: 1%{?dist}
+License: Apache
```
Which Apache license?  Also, you should use the short form from https://fedoraproject.org/wiki/Licensing:Main?rd=Licensing#Good_Licenses

mr-salty replied:
ASL 2.0 to match LICENSE (I did see that page before, but elsewhere I saw conflicting information about which format to use).

katzj commented on pkg/el/google-fluentd-catch-all-config.spec:
```diff
@@ -0,0 +1,40 @@
+License: Apache
+Group: System Environment/Daemons
+URL: https://cloud.google.com/logging/
+Requires: google-fluentd
```
You'll want a BuildArch: noarch so that it builds properly as a noarch package.

mr-salty replied:
done.

katzj commented on pkg/el/google-fluentd-catch-all-config.spec:
```diff
@@ -0,0 +1,40 @@
+from the system and third-party application packages.
+
+%files
+%config /etc/google-fluentd
```
You probably want

```
%dir /etc/google-fluentd
%config(noreplace) /etc/google-fluentd
```

to avoid overwriting things that users do

mr-salty replied:
done.

katzj commented on pkg/el/google-fluentd-catch-all-config.spec:
```diff
@@ -0,0 +1,40 @@
+Group: System Environment/Daemons
+URL: https://cloud.google.com/logging/
+Requires: google-fluentd
+
```
Worth adding an explicit build root too

```
BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-root
```

is the general recommendation

mr-salty replied:
done

katzj commented on pkg/el/google-fluentd-catch-all-config.spec:
```diff
@@ -0,0 +1,40 @@
+%config /etc/google-fluentd
+
+%clean
+# don't clean up the files here; we'll do it in the toplevel Makefile
```
%clean is more used for cleaning up the buildroot, so you probably want the usual (... or, if you leave it out, there's a default one in all modern versions of rpm)

mr-salty replied:
When I used the default one, it was cleaning up things I didn't want it to, which is why I did this... but I will try it again with buildroot.

mr-salty replied:
Yes, if I use the default %clean it does this:
- cd /usr/local/google/home/salty/rpmbuild/BUILD
- /bin/rm -rf /usr/local/google/home/salty/git/github/GoogleCloudPlatform/fluentd-catch-all-config/build/el/google-fluentd-catch-all-config-0.1/files

which is not what I want, since I then try to mv a file back which fails (and I also build a tarball out of that directory, although now that we're in the repo I may be able to drop tar support).  it's probably a consequence of the... "unorthodox" way I'm building the package - I populate that directory from the top level Makefile, not from the spec file, so I think that is why I don't want the "clean" step to happen here either.

katzj commented on pkg/el/google-fluentd-catch-all-config.spec:
```diff
@@ -0,0 +1,40 @@
+
+# post-install/un-install scripts to handle diverting google-fluentd.conf
+%triggerin -- google-fluentd
+if [ ! -h /etc/google-fluentd/google-fluentd.conf -o ! "`readlink /etc/google-fluentd/google-fluentd.conf`" = "/etc/google-fluentd/google-fluentd.conf.google" ] ; then
```
This seems like it's risky and likely to overwrite any changes a user made?

katzj commented on pkg/el/google-fluentd-catch-all-config.spec:
```diff
@@ -0,0 +1,40 @@
+%triggerin -- google-fluentd
+if [ ! -h /etc/google-fluentd/google-fluentd.conf -o ! "`readlink /etc/google-fluentd/google-fluentd.conf`" = "/etc/google-fluentd/google-fluentd.conf.google" ] ; then
+  if [ -e /etc/google-fluentd/google-fluentd.conf ] ; then
+    mv -f /etc/google-fluentd/google-fluentd.conf /etc/google-fluentd/google-fluentd.conf.orig
```
If we keep doing it, should probably do .rpmsave instead.

katzj commented on pkg/el/google-fluentd-catch-all-config.spec:
```diff
@@ -0,0 +1,40 @@
+  ln -s /etc/google-fluentd/google-fluentd.conf.google /etc/google-fluentd/google-fluentd.conf
+fi
+
+%triggerun -- ntp
```
Assuming you want this on google-fluentd instead

mr-salty replied:
yes, cut & paste. error - although, when I uninstalled the package it seemed to trigger this regardless.

katzj commented on pkg/el/google-fluentd-catch-all-config.spec:
```diff
@@ -0,0 +1,40 @@
+  mv -f /etc/google-fluentd/google-fluentd.conf.orig /etc/google-fluentd/google-fluentd.conf
+fi
+
+%triggerpostun -- ntp
```
Same

mr-salty replied:
done.

katzj commented:
Some small things but the bigger question is around what the triggers are being used for.  As a general thing, triggers are a great way to get yourself into a lot of trouble with rpms because they don't necessarily run when you expect them to

mr-salty commented:
What I'm trying to accomplish is moving google-fluentd.conf aside and replacing it when the package is uninstalled.  Without the triggers it saved the original to .rpmsave but did not restore it on uninstall.

For debian I used dpkg-divert to accomplish something similar.

As I mentioned in email, I got this idea from a presentation, but I don't have much experience - if there's a better way to accomplish this I am certainly open to changing it.
http://www.redhat.com/promo/summit/2010/presentations/summit/opensource-for-it-leaders/thurs/pwaterma-2-rpm/RPM-ifying-System-Configurations.pdf

mr-salty committed:
Incorporate comments from katzj

katzj commented:
That's generally a really bad idea with RPM.  There are all kinds of race conditions dependent on how you do the install and so you can't count on it working.

Is there a reason we can't push the changed config into the base package instead?

mr-salty commented:
Since google-fluentd is a fork, I wanted to minimize the changes I made there.
We also have users who don't use the catch-all configs, so keeping them separate simplifies their installation.
But, perhaps there is another approach that would work, let me do some investigation.

mr-salty commented on Makefile:
```diff
@@ -21,7 +24,14 @@ deb: populate-deb
 
+rpm: populate-el
+	mv ${EL_FILES_DIR}/google-fluentd.conf ${EL_FILES_DIR}/google-fluentd.conf.google
+	ls -l ${EL_FILES_DIR}
```
oops, this was something I added while debugging; I'll remove it.

mr-salty committed:
Merge branch 'master' into rpm-packaging

mr-salty committed:
Remove diversions from RPM packaging.

Since we no longer have google-fluentd.conf we don't have to
worry about trying to divert it, so this is much simpler.

mr-salty commented:
Ok, as we discussed offline, I pushed google-fluentd.conf into the main package to eliminate the need for diverting it here, so the result is much simpler.

I think I addressed all your prior comments, although there is still a bit of weirdness around "clean".

katzj commented:
%clean still looks weird but if it's okay when you're building it locally, it's fine until we get it running in a clean build system.

LGTM
